imap_data_access/file_validation.py

```diff
@@ -0,0 +1,226 @@
+"""Methods for managing and validating filenames and filepaths"""
+from __future__ import annotations
+
+import re
+from datetime import datetime
+from pathlib import Path
+
+import imap_data_access
+
+
+class ScienceFilepath:
+    class InvalidScienceFileError(Exception):
+        """Indicates a bad file type"""
+
+        pass
+
+    def __init__(self, filename: str | Path, data_dir: Path | None = None):
+        """Class to store filepath and file management methods for science files.
+
+        If you have an instance of this class, you can be confident you have a valid
+        science file and generate paths in the correct format.
+
+        Current filename convention:
+        <mission>_<instrument>_<datalevel>_<descriptor>_<startdate>_<enddate>
+        _<version>.<extension>
+
+        NOTE: There are no optional parameters anymore. All parameters are required.
+        <mission>: imap
+        <instrument>: idex, swe, swapi, hi-45, ultra-45 and etc.
+        <datalevel> : l1a, l1b, l1, l3a and etc.
+        <descriptor>: descriptor stores information specific to instrument. This is
+            decided by each instrument. For L0, "raw" is used.
+        <startdate>: startdate is the earliest date in the data. Format - YYYYMMDD
+        <enddate>: Some instrument and some data level requires to store date range.
+            If there is no end date, then startdate will be used as enddate as well.
+            Format - YYYYMMDD.
+        <version>: This stores software version and data version. Version format is
+            vxx-xx.
+
+        Parameters
+        ----------
+        filename : str | Path
+            Science data filename or file path.
+        data_dir : Path, optional
+            Optional higher directory level for the data, by default None
+        """
+        self.filename = Path(filename)
+        self.data_dir = data_dir
+
+        try:
+            split_filename = self.extract_filename_components(self.filename)
+        except ValueError as err:
+            raise self.InvalidScienceFileError(
+                f"Invalid filename. Expected file to match format: "
+                f"{imap_data_access.FILENAME_CONVENTION}"
+            ) from err
+
+        self.mission = split_filename["mission"]
+        self.instrument = split_filename["instrument"]
+        self.data_level = split_filename["datalevel"]
+        self.descriptor = split_filename["descriptor"]
+        self.startdate = split_filename["startdate"]
+        self.enddate = split_filename["enddate"]
+        self.version = split_filename["version"]
+        self.extension = split_filename["extension"]
+
+        self.error_message = self.validate_filename()
+        if self.error_message:
+            raise self.InvalidScienceFileError(f"{self.error_message}")
```

Comment:
Is there a reason to keep the error message on the class, or could we just raise from the end of the `validate_filename()` helper function directly instead?

```suggestion
        self.validate_filename()
```

Reply:
I could see people using validate_filename without wanting to throw an error - when I see `validate_something` I wouldn't assume it throws an error if it's invalid.

Reply:
I agree, but I also wouldn't expect an empty string or for an `error_message` attribute, so there are a few indirections here too.

It isn't a static method, so I think it'd only get used this one time, or are there places we'd want to use it after the fact too? Could make this private and call it `_raise_if_bad_filename()`.